doc/gnoi-reboot/gnoi_reboot.md

```diff
@@ -0,0 +1,252 @@
+
+
+
+## Revision
+
+
+Rev  | RevDate    | Author(s)   | Change Description
+---- | ---------- | ----------- | ------------------
+v0.1 | 09/22/2023 | John Hamrick (Google) | Initial Version
+
+
+
+## Scope
+
+_This document describes the high level design for SONiC gNOI Reboot RPC support._
+
+
+## Definitions/Abbreviations
+
+API: Application Programming Interface
+
+BE: Backend
+
+FE: Frontend
+
+gNOI: gRPC Network Operations Interface
+
+RPC: remote procedure call
+
+SAI: Switch Abstraction Interface
+
+SONiC: Software for Open Networking in the Cloud
+
+UMF: Unified Management Framework
+
+
+## Overview
+
+_Openconfig system.proto ([https://github.com/openconfig/gnoi/blob/main/system/system.proto](https://github.com/openconfig/gnoi/blob/main/system/system.proto)) defines reboot, cancel reboot and reboot status rpc’s within the system service. These rpc’s can be used to start, cancel and check status of reboot activities._
+
+
+## Requirements
+
+_NA_
+
+
+## Architecture Design
+
+The proposal is to create a Reboot Backend (BE) daemon in a new “Framework” container to handle a reboot request and forward it to the platform via DBUS.
+
+This daemon provides a location to implement application level reboot functionality before forwarding a reboot request to the platform (in the spirit of “[Docker to Host Communications HLD](https://github.com/sonic-net/SONiC/blob/master/doc/mgmt/Docker%20to%20Host%20communication.md)”.
+
+The reboot request contains a delay parameter, method and subcomponent fields. For subcomponents we can have support for:
+
+
+
+* system reboot (method = warm, cold, fast, etc)
+* optic reboot
+* container reboot
+* etc
+
+The immediate short term use case for the reboot backend daemon is to support system warm boot. During warmboot applications will be requested and confirmed to:
+
+
+
+* freeze/quiesce
+* verify state
+* checkpoint (save) state before shutdown 
+* reconcile after boot
+* verify state
+* unfreeze
+
+
+
+
+
+![drawing](./images/image3.png)
+
+The controller can send a reboot request to the UMF gNOI Reboot Front End. This request will be forwarded to a Reboot Back End daemon executing in a container. The reboot back end daemon will then request host/platform level reboot actions via DBUS and a reboot sonic host service.
+
+### Framework Container
+
+There is no obvious container home for a reboot backend daemon: SWSS and PMON were considered. SWSS has the primary role of processing CONFIG_DB inputs (or kernel events) to APP_DB and ASIC_DB outputs. PMON, by name, exists to monitor the platform. Further, there are plans for a state verification daemon which also needs a home.
+
+The Framework container is envisioned to provide a home for foundational services that interact with multiple containers/applications.
+
+State verification daemon periodically (or by request) asks applications to verify their state. For example:
+
+
+
+* SWSS Cfgmgr confirms CONFIG DB matches APPL DB.
```

Comment:
>Cfgmgr [](http://example.com/codeflow?start=7&length=6)

Why we need "State verification daemon"? Cfgmgr could check matching and write to StateDB, and Syncd could check matching and write to StateDB.

Reply:
This is a bit of tangent from reboot HLD, but I mentioned it:

State verification daemon periodically and by request asks other system entities to verify their state.
- the daemon aggregates results to report a "system" state
- anyone can ask the daemon to verify and report state for the entire system

Why? I'll leave that to a forthcoming state verification HLD (authored by someone else).

Reply:
If this is not part of your design, could you remove this section, or link to anther HLD?